File: docx_parsers/document/margins_parser.py
from lxml import etree
from typing import Optional
from docx_parsers.helpers.common_helpers import extract_element, extract_attribute
from docx_parsers.utils import convert_twips_to_points
from docx_parsers.models.document_models import DocMargins

class MarginsParser:
    @staticmethod
    def parse(sectPr: Optional[etree.Element]) -> Optional[DocMargins]:
        """
        Parses margins from the given section properties XML element.

        Args:
            sectPr (Optional[etree.Element]): The section properties XML element.

        Returns:
            DocMargins: The parsed margins.
        """
        pgMar = extract_element(sectPr, ".//w:pgMar")
        if pgMar is not None:
            top = extract_attribute(pgMar, 'top')
            right = extract_attribute(pgMar, 'right') or extract_attribute(pgMar, 'end')
            bottom = extract_attribute(pgMar, 'bottom')
            left = extract_attribute(pgMar, 'left') or extract_attribute(pgMar, 'start')
            header = extract_attribute(pgMar, 'header')
            footer = extract_attribute(pgMar, 'footer')
            gutter = extract_attribute(pgMar, 'gutter')

            return DocMargins(
                top_pt=convert_twips_to_points(int(top)) if top is not None else None,
                right_pt=convert_twips_to_points(int(right)) if right is not None else None,
                bottom_pt=convert_twips_to_points(int(bottom)) if bottom is not None else None,
                left_pt=convert_twips_to_points(int(left)) if left is not None else None,
                header_pt=convert_twips_to_points(int(header)) if header is not None else None,
                footer_pt=convert_twips_to_points(int(footer)) if footer is not None else None,
                gutter_pt=convert_twips_to_points(int(gutter)) if gutter is not None else None
            )
        return None

File: docx_parser_converter_ts/src/docx_parsers/document/marginsParser.ts
import type { DocMargins } from '../models/documentModels';
import { extractElement, extractAttribute, safeInt } from '../helpers/commonHelpers';
import { convertTwipsToPoints } from '../utils';

/**
 * Parses the margin properties of a section in a docx document.
 *
 * This class contains methods to parse the margin properties from the
 * section properties (sectPr) element of a docx document. The margin
 * properties are essential for understanding the layout of the document.
 */
export class MarginsParser {
    /**
     * Parses margins from the given section properties XML element.
     *
     * @param sectPr - The section properties XML element (Element or null).
     *                 This element can contain margin properties which define the
     *                 top, right, bottom, left, header, footer, and gutter margins
     *                 of the section.
     * @returns The parsed margins as a DocMargins object, or null if not found or sectPr is null.
     *
     * @example
     * The following is an example of the section properties with margins
     * in a document.xml file:
     *
     * ```xml
     * <w:sectPr xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main">
     *     <w:pgMar w:left="1134" w:right="1134" w:gutter="0" w:header="0"
     *              w:top="1134" w:footer="0" w:bottom="1134"/>
     * </w:sectPr>
     * ```
     */
    public static parse(sectPr: Element | null): DocMargins | null {
        if (!sectPr) {
            return null;
        }

        const pgMar = extractElement(sectPr, ".//w:pgMar");
        if (pgMar) {
            const topStr = extractAttribute(pgMar, 'top');
            const rightStr = extractAttribute(pgMar, 'right') || extractAttribute(pgMar, 'end');
            const bottomStr = extractAttribute(pgMar, 'bottom');
            const leftStr = extractAttribute(pgMar, 'left') || extractAttribute(pgMar, 'start');
            const headerStr = extractAttribute(pgMar, 'header');
            const footerStr = extractAttribute(pgMar, 'footer');
            const gutterStr = extractAttribute(pgMar, 'gutter');

            const topTwips = safeInt(topStr);
            const rightTwips = safeInt(rightStr);
            const bottomTwips = safeInt(bottomStr);
            const leftTwips = safeInt(leftStr);
            const headerTwips = safeInt(headerStr);
            const footerTwips = safeInt(footerStr);
            const gutterTwips = safeInt(gutterStr);

            const margins: DocMargins = {};

            if (topTwips !== null) margins.topPt = convertTwipsToPoints(topTwips);
            if (rightTwips !== null) margins.rightPt = convertTwipsToPoints(rightTwips);
            if (bottomTwips !== null) margins.bottomPt = convertTwipsToPoints(bottomTwips);
            if (leftTwips !== null) margins.leftPt = convertTwipsToPoints(leftTwips);
            if (headerTwips !== null) margins.headerPt = convertTwipsToPoints(headerTwips);
            if (footerTwips !== null) margins.footerPt = convertTwipsToPoints(footerTwips);
            if (gutterTwips !== null) margins.gutterPt = convertTwipsToPoints(gutterTwips);
            
            // Return margins only if at least one property was parsed
            if (Object.keys(margins).length > 0) {
                return margins;
            }
        }
        return null;
    }
}
